File: practica_6/AlgebraVectorialBidimensional/main.py
import math

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return f"Vector{{{self.x} {self.y}}}"

    def suma(self, v):
        return Vector(self.x + v.x, self.y + v.y)

    def resta(self, v):
        return Vector(self.x - v.x, self.y - v.y)

    def producto_escalar(self, escalar):
        return Vector(self.x * escalar, self.y * escalar)

    def producto_punto(self, v):
        return self.x * v.x + self.y * v.y

    def producto_cruz(self, v):
        return self.x * v.y - self.y * v.x

    def magnitud(self):
        return math.sqrt(self.x * self.x + self.y * self.y)

    def angulo(self, v):
        return math.acos(self.producto_punto(v) / (self.magnitud() * v.magnitud()))

    def es_perpendicular(self, v):
        return "Son perpendiculares" if self.producto_punto(v) == 0 else "No son perpendiculares"

    def es_paralelo(self, v):
        return "Son paralelos" if self.producto_cruz(v) == 0 else "No son paralelos"

    def proyeccion(self, v):
        return v.producto_escalar(self.producto_punto(v) / (v.magnitud() * v.magnitud()))

    def componente(self, v):
        return self.producto_punto(v) / v.magnitud()

# Ejemplo de uso
if __name__ == "__main__":
    v1 = Vector(-2, 4)
    v2 = Vector(1, -2)

    v3 = Vector(3, 2)
    v4 = Vector(-2, 3)

    print(v1.es_perpendicular(v2))
    print(v1.es_paralelo(v2))
    print(v1.proyeccion(v2))
    print("El componente de es", v1.componente(v2))

    print()

    print(v3.es_perpendicular(v4))
    print(v3.es_paralelo(v4))
    print(v3.proyeccion(v4))
    print("El componente de es", v3.componente(v4))

File: practica_6/AlgebraVectorialBidimensional/Main.java
class Vector {
  private double x;
  private double y;

  public Vector(double x, double y) {
    this.x = x;
    this.y = y;
  }

  @Override
  public String toString() {
    return "Vector{" + x + " " + y + "}";
  }

  public Vector suma(Vector v) {
    return new Vector(this.x + v.x, this.y + v.y);
  }

  public Vector resta(Vector v) {
    return new Vector(this.x - v.x, this.y - v.y);
  }

  public Vector productoEscalar(double escalar) {
    return new Vector(this.x * escalar, this.y * escalar);
  }

  public double productoPunto(Vector v) {
    return this.x * v.x + this.y * v.y;
  }

  public double productoCruz(Vector v) {
    return this.x * v.y - this.y * v.x;
  }

  public double magnitud() {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  public double angulo(Vector v) {
    return Math.acos(this.productoPunto(v) / (this.magnitud() * v.magnitud()));
  }

  public String esPerpendicular(Vector v) {
    return this.productoPunto(v) == 0 ? "Son perpendiculares" : "No son perpendiculares";
  }

  public String esParalelo(Vector v) {
    return this.productoCruz(v) == 0 ? "Son paralelos" : "No son paralelos";
  }

  public Vector proyeccion(Vector v) {
    return v.productoEscalar(this.productoPunto(v) / (v.magnitud() * v.magnitud()));
  }

  public double componente(Vector v) {
    return this.productoPunto(v) / v.magnitud();
  }
}

public class Main {
  public static void main(String[] args) {
    Vector v1 = new Vector(-2, 4);
    Vector v2 = new Vector(1, -2);
    
    Vector v3 = new Vector(3,2);
    Vector v4 = new Vector(-2,3);

    System.out.println(v1.esPerpendicular(v2));
    System.out.println(v1.esParalelo(v2));
    System.out.println(v1.proyeccion(v2));
    System.out.println("El componente de es " + v1.componente(v2));
    
    System.out.println();
    
    System.out.println(v3.esPerpendicular(v4));
    System.out.println(v3.esParalelo(v4));
    System.out.println(v3.proyeccion(v4));
    System.out.println("El componente de es " + v3.componente(v4));
  }
}
